crates/primitives/src/result.rs

```diff
@@ -1,26 +1,35 @@
-use crate::{Address, Bytes, EvmState, Log, U256};
-use core::fmt;
+use derive_where::derive_where;
+
+use crate::{Address, Bytes, ChainSpec, EvmState, Log, TransactionValidation, U256};
+use core::fmt::{self, Debug};
 use std::{boxed::Box, string::String, vec::Vec};
 
 /// Result of EVM execution.
-pub type EVMResult<DBError> = EVMResultGeneric<ResultAndState, DBError>;
+pub type EVMResult<ChainSpecT, DBError> =
+    EVMResultGeneric<ResultAndState<ChainSpecT>, ChainSpecT, DBError>;
 
 /// Generic result of EVM execution. Used to represent error and generic output.
-pub type EVMResultGeneric<T, DBError> = core::result::Result<T, EVMError<DBError>>;
+pub type EVMResultGeneric<T, ChainSpecT, DBError> =
+    core::result::Result<T, EVMErrorForChain<DBError, ChainSpecT>>;
+
+/// EVM error type for a specific chain.
+pub type EVMErrorForChain<DBError, ChainSpecT> = EVMError<
+    DBError,
+    <<ChainSpecT as ChainSpec>::Transaction as TransactionValidation>::ValidationError,
+>;
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-pub struct ResultAndState {
+pub struct ResultAndState<ChainSpecT: ChainSpec> {
     /// Status of execution
-    pub result: ExecutionResult,
+    pub result: ExecutionResult<ChainSpecT>,
     /// State that got updated
     pub state: EvmState,
 }
 
 /// Result of a transaction execution.
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
```

Comment:
Serde's derive macro allows you to specify trait bounds. That way the derivation of the `Serialize` and `Deserialize` traits can be conditional on types with the `EvmWiring::Hardfork` type, depending on whether that type is (de)serializable.

Imo, it should be solved that way instead of always requiring the hardfork type to be serializable. We should only require trait bounds that we need internally

Reply:
Agreeing with this, but It becomes hard to custom implement serde support.

But found another solution, to use only `HaltReasonTrait` there: https://github.com/bluealloy/revm/blob/rakita/db_ext_wiring_temp/crates/primitives/src/result.rs#L33 in this case serde macro will handle it.